File: Wire-iOS/Sources/Helpers/UIColor+Mixing.m
//
// Wire
// Copyright (C) 2016 Wire Swiss GmbH
// 
// This program is free software: you can redistribute it and/or modify
// it under the terms of the GNU General Public License as published by
// the Free Software Foundation, either version 3 of the License, or
// (at your option) any later version.
// 
// This program is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
// GNU General Public License for more details.
// 
// You should have received a copy of the GNU General Public License
// along with this program. If not, see http://www.gnu.org/licenses/.
// 


#import "UIColor+Mixing.h"


static CGFloat mix(CGFloat value0, CGFloat value1, double progress)
{
    return (CGFloat) (value0 * (1 - progress) + value1 * progress);
}



@implementation UIColor (Mixing)


- (UIColor *)mix:(UIColor *)color amount:(double)progress
{
    CGFloat red0, green0, blue0, alpha0,
            red1, green1, blue1, alpha1;
    [self getRed:&red0 green:&green0 blue:&blue0 alpha:&alpha0];
    [color getRed:&red1 green:&green1 blue:&blue1 alpha:&alpha1];
    CGFloat red = mix(red0, red1, progress);
    CGFloat green = mix(green0, green1, progress);
    CGFloat blue = mix(blue0, blue1, progress);
    CGFloat alpha = mix(alpha0, alpha1, progress);

    return [UIColor colorWithRed:red green:green blue:blue alpha:alpha];
}

- (UIColor *)removeAlphaByBlendingWithColor:(UIColor *)color
{
    CGFloat red0, green0, blue0, alpha0;
    [self getRed:&red0 green:&green0 blue:&blue0 alpha:&alpha0];
    
    CGFloat red1, green1, blue1;
    [color getRed:&red1 green:&green1 blue:&blue1 alpha:nil];
    
    CGFloat red = mix(red1, red0, alpha0);
    CGFloat green = mix(green1, green0, alpha0);
    CGFloat blue = mix(blue1, blue0, alpha0);
    
    return [UIColor colorWithRed:red green:green blue:blue alpha:1];
}

+ (UIColor *)wr_colorFromString:(NSString *)string
{
    NSScanner *scanner = [NSScanner scannerWithString:string];
    
    scanner.charactersToBeSkipped = [NSCharacterSet characterSetWithCharactersInString:@"rgba(), "];
    
    float r = 0, g = 0, b = 0, a = 1;
    [scanner scanFloat:&r];
    [scanner scanFloat:&g];
    [scanner scanFloat:&b];
    [scanner scanFloat:&a];
    
    if (scanner.atEnd) {
        return [UIColor colorWithRed:r / 255 green:g / 255 blue:b / 255 alpha:a];
    } else {
        return nil;
    }
}

@end

File: wire-ios-utilities/Source/Public/UIColor/UIColor+Mixing.swift
//
// Wire
// Copyright (C) 2019 Wire Swiss GmbH
//
// This program is free software: you can redistribute it and/or modify
// it under the terms of the GNU General Public License as published by
// the Free Software Foundation, either version 3 of the License, or
// (at your option) any later version.
//
// This program is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
// GNU General Public License for more details.
//
// You should have received a copy of the GNU General Public License
// along with this program. If not, see http://www.gnu.org/licenses/.
//

import UIKit

extension UIColor {

    fileprivate func mix(value0: CGFloat, value1: CGFloat, progress: CGFloat) -> CGFloat {
        return value0 * (1 - progress) + value1 * progress
    }

    /// Pass in amount of 0 for self, 1 is the other color
    ///
    /// - Parameters:
    ///   - color: other color to mix
    ///   - progress: amount of other color
    /// - Returns: the mixed color
    @objc
    public func mix(_ color: UIColor, amount progress: CGFloat) -> UIColor {
        
        let component0 = components
        let component1 = color.components

        let red = mix(value0: component0.red, value1: component1.red, progress: progress)
        let green = mix(value0: component0.green, value1: component1.green, progress: progress)
        let blue = mix(value0: component0.blue, value1: component1.blue, progress: progress)
        let alpha = mix(value0: component0.alpha, value1: component1.alpha, progress: progress)
        
        return UIColor(red: red, green: green, blue: blue, alpha: alpha)
    }
    
    public func removeAlphaByBlending(with color: UIColor) -> UIColor {
        let component0 = components
        let component1 = color.components
        let alpha0 = component0.alpha
        
        let red = mix(value0: component1.red, value1: component0.red, progress: alpha0)
        let green = mix(value0: component1.green, value1: component0.green, progress: alpha0)
        let blue = mix(value0: component1.blue, value1: component0.blue, progress: alpha0)
        
        return UIColor(red: red, green: green, blue: blue, alpha: 1)
    }
}
